File: front/src/services/common.service.ts
import { SecurityChargeCode } from '../app/shared/model/common/security-charge-code.interface';
import {Injectable} from '@angular/core';
import {Response} from '@angular/http';
import {HttpInterceptorService} from '@covalent/http';
import {Observable} from 'rxjs';
import {environment} from '../environments/environment';
import {CohortCode} from '../app/shared/model/common/cohort-code.interface';
import {FacultyCode} from '../app/shared/model/common/faculty-code.interface';
import {CountryCode} from '../app/shared/model/common/country-code.interface';
import {ProgramCode} from '../app/shared/model/common/program-code.interface';
import {StudyMode} from '../app/shared/model/common/study-mode.interface';
import {StateCode} from '../app/shared/model/common/state-code.interface';
import {BankCode} from '../app/shared/model/common/bank-code.interface';
import {ResidencyCode} from '../app/shared/model/common/residency-code.interface';
import {TaxCode} from '../app/shared/model/common/tax-code.interface';

@Injectable()
export class CommonService {

  private COMMON_API: string = environment.endpoint + '/api/common';

  constructor(private _http: HttpInterceptorService) {
  }

  // ====================================================================================================
  // BANK CODES
  // ====================================================================================================

  findBankCodes(): Observable<BankCode[]> {
    console.log('findBankCodes()');
    return this._http.get(this.COMMON_API + '/bankCodes')
      .map((res: Response) => <BankCode[]>res.json());
  }

  findBankCodeByCode(code: string): Observable<BankCode> {
    console.log('findBankCodeByCode');
    return this._http.get(this.COMMON_API + '/bankCodes/' + code)
      .map((res: Response) => <BankCode>res.json());
  }

  saveBankCode(code: BankCode) {
    return this._http.post(this.COMMON_API + '/bankCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateBankCode(code: BankCode) {
    return this._http.put(this.COMMON_API + '/bankCodes/' + code.code, JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeBankCode(code: BankCode) {
    return this._http.delete(this.COMMON_API + '/bankCodes/' + code.code)
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // COHORT CODES
  // ====================================================================================================

  findCohortCodes(): Observable<CohortCode[]> {
    console.log('findCohortCodes');
    return this._http.get(this.COMMON_API + '/cohortCodes')
      .map((res: Response) => <CohortCode[]>res.json());
  }

  findCohortCodeByCode(code: string): Observable<CohortCode> {
    console.log('findCohortCodeByCode');
    return this._http.get(this.COMMON_API + '/cohortCodes/' + code)
      .map((res: Response) => <CohortCode>res.json());
  }

  saveCohortCode(code: CohortCode) {
    return this._http.post(this.COMMON_API + '/cohortCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeCohortCode(code: CohortCode) {
    return this._http.delete(this.COMMON_API + '/cohortCodes/' + code.code)
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateCohortCode(code: CohortCode) {
    return this._http.put(this.COMMON_API + '/cohortCodes/' + code.code, JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // FACULTY CODES
  // ====================================================================================================

  findFacultyCodes(): Observable<FacultyCode[]> {
    console.log('findFacultyCodes');
    return this._http.get(this.COMMON_API + '/facultyCodes')
      .map((res: Response) => <FacultyCode[]>res.json());
  }

  findFacultyCodeByCode(code: string): Observable<FacultyCode> {
    console.log('findFacultyCodeByCode');
    return this._http.get(this.COMMON_API + '/facultyCodes/' + code)
      .map((res: Response) => <FacultyCode>res.json());
  }

  findProgramCodesByFacultyCode(facultyCode: FacultyCode): Observable<ProgramCode[]> {
    console.log('findProgramCodesByFacultyCode');
    return this._http.get(this.COMMON_API + '/facultyCodes/' + facultyCode.code + '/programCodes')
      .map((res: Response) => <ProgramCode[]>res.json());
  }

  saveFacultyCode(code: FacultyCode) {
    return this._http.post(this.COMMON_API + '/facultyCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }


  // ====================================================================================================
  // PROGRAM CODES
  // ====================================================================================================

  findProgramCodes(): Observable<ProgramCode[]> {
    console.log('findProgramCodes');
    return this._http.get(this.COMMON_API + '/programCodes')
      .map((res: Response) => <ProgramCode[]>res.json());
  }

  findProgramCodeByCode(code: string): Observable<ProgramCode> {
    console.log('findProgramCodeByCode');
    return this._http.get(this.COMMON_API + '/programCodes/' + code)
      .map((res: Response) => <ProgramCode>res.json());
  }

  saveProgramCode(code: ProgramCode) {
    return this._http.post(this.COMMON_API + '/programCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeProgramCode(code: ProgramCode) {
    return this._http.delete(this.COMMON_API + '/programCodes/' + code.code)
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // RESIDENCY CODES
  // ====================================================================================================

  findResidencyCodes(): Observable<ResidencyCode[]> {
    console.log('findResidencyCodes');
    return this._http.get(this.COMMON_API + '/residencyCodes')
      .map((res: Response) => <ResidencyCode[]>res.json());
  }

  saveResidencyCode(code: ResidencyCode) {
    return this._http.post(this.COMMON_API + '/residencyCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateResidencyCode(code: ResidencyCode) {
    return this._http.put(this.COMMON_API + '/residencyCodes/' + code.code, JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeResidencyCode(code: ResidencyCode) {
    return this._http.delete(this.COMMON_API + '/residencyCodes/' + code.code)
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // STUDY MODE
  // ====================================================================================================

  findStudyModes(): Observable<StudyMode[]> {
    console.log('findStudyModes');
    return this._http.get(this.COMMON_API + '/studyModes')
      .map((res: Response) => <StudyMode[]>res.json());
  }

  findStudyModeByCode(code: string): Observable<StudyMode> {
    console.log('findStudyModeByCode');
    return this._http.get(this.COMMON_API + '/studyModes/' + code)
      .map((res: Response) => <StudyMode>res.json());
  }

  saveStudyMode(code: StudyMode) {
    return this._http.post(this.COMMON_API + '/studyModes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

// ====================================================================================================
// COUNTRY CODES
// ====================================================================================================

  findCountryCodes(): Observable<CountryCode[]> {
    console.log('findCountryCodes()');
    return this._http.get(this.COMMON_API + '/countryCodes')
      .map((res: Response) => <CountryCode[]>res.json());
  }

  findCountryCodeByCode(code: string): Observable<CountryCode> {
    console.log('findCountryCodeByCode');
    return this._http.get(this.COMMON_API + '/countryCodes/' + code)
      .map((res: Response) => <CountryCode>res.json());
  }

// ====================================================================================================
// STATE CODES
// ====================================================================================================

  findStateCodes(): Observable<StateCode[]> {
    console.log('findStateCodes()');
    return this._http.get(this.COMMON_API + '/stateCodes')
      .map((res: Response) => <StateCode[]>res.json());
  }

  findStateCodeByCode(code: string): Observable<StateCode> {
    console.log('findStateCodeByCode');
    return this._http.get(this.COMMON_API + '/stateCodes/' + code)
      .map((res: Response) => <StateCode>res.json());
  }

  // ====================================================================================================
  // TAX CODES
  // ====================================================================================================

  findTaxCodes(): Observable<TaxCode[]> {
    console.log('findTaxCodes');
    return this._http.get(this.COMMON_API + '/taxCodes')
      .map((res: Response) => <TaxCode[]>res.json());
  }

  saveTaxCode(code: TaxCode) {
    return this._http.post(this.COMMON_API + '/taxCodes', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateTaxCode(code: TaxCode) {
    return this._http.put(this.COMMON_API + '/taxCodes/' + code.code, JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeTaxCode(code: TaxCode) {
    return this._http.delete(this.COMMON_API + '/taxCodes/' + code.code)
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // SECURITY Charge CODES
  // ====================================================================================================

  findSecurityChargeCodes(): Observable<SecurityChargeCode[]> {
    console.log('findSecurityChargeCodes');
    return this._http.get(this.COMMON_API + '/securityChargeCodes')
      .map((res: Response) => <SecurityChargeCode[]>res.json());
  }

  findSecurityChargeCodesByActive(): Observable<SecurityChargeCode[]> {
    console.log('findSecurityChargeCodesActive');
    return this._http.get(this.COMMON_API + '/securityChargeCodesActive')
      .map((res: Response) => <SecurityChargeCode[]>res.json());
  }

  saveSecurityChargeCode(code: SecurityChargeCode) {
    return this._http.post(this.COMMON_API + '/securityChargeCode', JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateSecurityChargeCode(code: SecurityChargeCode) {
    return this._http.put(this.COMMON_API + '/securityChargeCode/' + code.id, JSON.stringify(code))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeSecurityChargeCode(code: SecurityChargeCode) {
    return this._http.delete(this.COMMON_API + '/securityChargeCode/' + code.id)
      .flatMap((res: Response) => Observable.of(res.text()));
  }
}

File: back/src/main/java/my/edu/umk/pams/account/common/service/CommonService.java
package my.edu.umk.pams.account.common.service;

import my.edu.umk.pams.account.common.model.*;

import java.util.List;

public interface CommonService {

    //====================================================================================================
    // COUNTRY CODE
    //====================================================================================================

    AcCountryCode findCountryCodeById(Long id);

    AcCountryCode findCountryCodeByCode(String code);

    List<AcCountryCode> findCountryCodes(String filter, Integer offset, Integer limit);

    Integer countCountryCode(String filter);

    void saveCountryCode(AcCountryCode countryCode);

    void updateCountryCode(AcCountryCode countryCode);

    void removeCountryCode(AcCountryCode countryCode);


    //====================================================================================================
    // STATE CODE
    //====================================================================================================

    AcStateCode findStateCodeById(Long id);

    AcStateCode findStateCodeByCode(String code);

    List<AcStateCode> findStateCodes(Integer offset, Integer limit);

    List<AcStateCode> findStateCodes(String filter, Integer offset, Integer limit);

    List<AcStateCode> findStateCodes(AcCountryCode countryCode, Integer offset, Integer limit);

    List<AcStateCode> findStateCodes(String filter, AcCountryCode countryCode, Integer offset, Integer limit);

    Integer countStateCode();

    Integer countStateCode(String filter);

    Integer countStateCode(AcCountryCode countryCode);

    Integer countStateCode(String filter, AcCountryCode countryCode);

    void saveStateCode(AcStateCode stateCode);

    void updateStateCode(AcStateCode stateCode);

    void removeStateCode(AcStateCode stateCode);

    //====================================================================================================
    // DISTRICT CODE
    //====================================================================================================

    AcDistrictCode findDistrictCodeById(Long id);

    AcDistrictCode findDistrictCodeByCode(String code);

    List<AcDistrictCode> findDistrictCodes(AcStateCode stateCod, Integer offset, Integer limit);

    List<AcDistrictCode> findDistrictCodes(String filter, AcStateCode stateCode, Integer offset, Integer limit);

    Integer countDistrictCode(AcStateCode stateCode);

    Integer countDistrictCode(String filter, AcStateCode stateCode);

    void saveDistrictCode(AcDistrictCode districtCode);

    void updateDistrictCode(AcDistrictCode districtCode);

    void removeDistrictCode(AcDistrictCode districtCode);

    //====================================================================================================
    // CITY CODE
    //====================================================================================================

    AcCityCode findCityCodeById(Long id);

    AcCityCode findCityCodeByCode(String code);

    List<AcCityCode> findCityCodes(AcStateCode stateCode, Integer offset, Integer limit);

    List<AcCityCode> findCityCodes(String filter, AcStateCode stateCode, Integer offset, Integer limit);

    Integer countCityCode(AcStateCode stateCode);

    Integer countCityCode(String filter, AcStateCode stateCode);

    void saveCityCode(AcCityCode cityCode);

    void updateCityCode(AcCityCode cityCode);

    void removeCityCode(AcCityCode cityCode);

    //====================================================================================================
    // FACULTY CODE
    //====================================================================================================

    AcFacultyCode findFacultyCodeById(Long id);

    AcFacultyCode findFacultyCodeByCode(String code);

    List<AcFacultyCode> findFacultyCodes();
    
    List<AcFacultyCode> findFacultyCodes(String filter, Integer offset, Integer limit);

    Integer countFacultyCode();

    Integer countFacultyCode(String filter);

    void saveFacultyCode(AcFacultyCode facultyCode);

    void updateFacultyCode(AcFacultyCode facultyCode);

    void removeFacultyCode(AcFacultyCode facultyCode);

    //====================================================================================================
    // STUDY CENTER CODE
    //====================================================================================================

    AcStudyCenterCode findStudyCenterCodeById(Long id);

    AcStudyCenterCode findStudyCenterCodeByCode(String code);

    List<AcStudyCenterCode> findStudyCenterCodes();

    List<AcStudyCenterCode> findStudyCenterCodes(String filter, Integer offset, Integer limit);

    Integer countStudyCenterCode();

    Integer countStudyCenterCode(String filter);

    void saveStudyCenterCode(AcStudyCenterCode studyCenterCode);

    void updateStudyCenterCode(AcStudyCenterCode studyCenterCode);

    void removeStudyCenterCode(AcStudyCenterCode studyCenterCode);

    //====================================================================================================
    // BANK CODE
    //====================================================================================================

    AcBankCode findBankCodeById(Long id);

    AcBankCode findBankCodeByCode(String code);

    List<AcBankCode> findBankCodes();

    List<AcBankCode> findBankCodes(String filter, Integer offset, Integer limit);

    Integer countBankCode();

    Integer countBankCode(String filter);

    void saveBankCode(AcBankCode bankCode);

    void updateBankCode(AcBankCode bankCode);

    void removeBankCode(AcBankCode bankCode);


    //====================================================================================================
    // PROGRAM CODE
    //====================================================================================================

    AcProgramCode findProgramCodeById(Long id);

    AcProgramCode findProgramCodeByCode(String code);

    AcProgramCode findProgramCodeByUpuCode(String upuCode);

    List<AcProgramCode> findProgramCodes(AcFacultyCode facultyCode);

    List<AcProgramCode> findProgramCodes(String filter, Integer offset, Integer limit);

    Integer countProgramCode();

    Integer countProgramCode(String filter);

    void saveProgramCode(AcProgramCode programCode);

    void updateProgramCode(AcProgramCode programCode);

    void removeProgramCode(AcProgramCode programCode);


    //====================================================================================================
    // NATIONALITY CODE
    //====================================================================================================

    AcNationalityCode findNationalityCodeById(Long id);

    AcNationalityCode findNationalityCodeByCode(String code);

    List<AcNationalityCode> findNationalityCodes();

    List<AcNationalityCode> findNationalityCodes(String filter, Integer offset, Integer limit);

    Integer countNationalityCode();

    Integer countNationalityCode(String filter);

    void saveNationalityCode(AcNationalityCode nationalityCode);

    void updateNationalityCode(AcNationalityCode nationalityCode);

    void removeNationalityCode(AcNationalityCode nationalityCode);


    //====================================================================================================
    // RESIDENCY CODE
    //====================================================================================================

    AcResidencyCode findResidencyCodeById(Long id);

    AcResidencyCode findResidencyCodeByCode(String code);

    List<AcResidencyCode> findResidencyCodes();

    List<AcResidencyCode> findResidencyCodes(String filter, Integer offset, Integer limit);

    Integer countResidencyCode();

    Integer countResidencyCode(String filter);

    void saveResidencyCode(AcResidencyCode residencyCode);

    void updateResidencyCode(AcResidencyCode residencyCode);

    void removeResidencyCode(AcResidencyCode residencyCode);


    //====================================================================================================
    // COHORT CODE
    //====================================================================================================

    AcCohortCode findCohortCodeById(Long id);

    AcCohortCode findCohortCodeByCode(String code);

    List<AcCohortCode> findCohortCodes();

    List<AcCohortCode> findCohortCodes(Integer offset, Integer limit);

    List<AcCohortCode> findCohortCodes(String filter, Integer offset, Integer limit);

    Integer countCohortCode();

    Integer countCohortCode(String filter);

    void saveCohortCode(AcCohortCode cohortCode);

    void updateCohortCode(AcCohortCode cohortCode);

    void removeCohortCode(AcCohortCode cohortCode);

    // ====================================================================================================
    // STUDY MODE
    // ====================================================================================================

    AcStudyMode findStudyModeById(Long id);

    AcStudyMode findStudyModeByCode(String code);

    List<AcStudyMode> findStudyModes();

    List<AcStudyMode> findStudyModes(String filter, Integer offset, Integer limit);

    Integer countStudyMode();

    Integer countStudyMode(String filter);

    void saveStudyMode(AcStudyMode studyMode);

    void updateStudyMode(AcStudyMode studyMode);

    void removeStudyMode(AcStudyMode studyMode);

    //====================================================================================================
    // TAX CODE
    //====================================================================================================

    AcTaxCode findTaxCodeById(Long id);

    AcTaxCode findTaxCodeByCode(String code);

    List<AcTaxCode> findTaxCodes();

    List<AcTaxCode> findTaxCodes(String filter, Integer offset, Integer limit);

    Integer countTaxCode();

    Integer countTaxCode(String filter);

    void saveTaxCode(AcTaxCode taxCode);

    void updateTaxCode(AcTaxCode taxCode);

    void removeTaxCode(AcTaxCode taxCode);
    
    //====================================================================================================
    // SECURITY Charge CODE 
    //====================================================================================================
    
    AcSecurityChargeCode findSecurityChargeCodeById(Long id);

    AcSecurityChargeCode findSecurityChargeCodeByCode(String code);

    List<AcSecurityChargeCode> findSecurityChargeCodes();

    List<AcSecurityChargeCode> findSecurityChargeCodes(String filter, Integer offset, Integer limit);

    Integer countSecurityChargeCode();

    Integer countSecurityChargeCode(String filter);

    void saveSecurityChargeCode(AcSecurityChargeCode securityChargeCode);

    void updateSecurityChargeCode(AcSecurityChargeCode securityChargeCode);

    void removeSecurityChargeCode(AcSecurityChargeCode securityChargeCode);

	void calculateSecurityChargeCodeNetAmount(AcSecurityChargeCode securityChargeCode);

	List<AcSecurityChargeCode> findSecurityChargeCodesByActive(Boolean active);




    
}
